src/Text/Pandoc/Writers/Shared.hs

```diff
@@ -166,17 +171,48 @@ getLang opts meta =
 
 -- | Produce an HTML tag with the given pandoc attributes.
 tagWithAttrs :: HasChars a => Text -> Attr -> Doc a
-tagWithAttrs tag (ident,classes,kvs) = hsep
-  ["<" <> text (T.unpack tag)
-  ,if T.null ident
+tagWithAttrs tag attr = "<" <> text (T.unpack tag) <> (htmlAttrs attr) <> ">"
```

Comment:
I'm pretty sure `literal tag` will do the same thing and avoid the unpack. (But I know this was in the original code, so no worries if you don't want to change it.)

Reply:
thanks, will do!

Reply:
I don't understand it, it doesn't seem to compile:

using `text (literal tag)` instead of `text (T.Unpack tag)`:
```
src/Text/Pandoc/Writers/Shared.hs:174:38: error:
    • Couldn't match type: Doc Text
                     with: [Char]
      Expected: String
        Actual: Doc Text
    • In the first argument of ‘text’, namely ‘(literal tag)’
      In the first argument of ‘(<>)’, namely ‘text (literal tag)’
      In the second argument of ‘(<>)’, namely
        ‘text (literal tag) <> (htmlAttrs attr) <> ">"’
    |
174 | tagWithAttrs tag attr = "<" <> text (literal tag) <> (htmlAttrs attr) <> ">"
    |                                      ^^^^^^^^^^^
cabal: Failed to build pandoc-3.0 (which is required by exe:pandoc from
pandoc-cli-0.1).
```

Using `(literal tag)` instead of `text (T.unpack tag)`: 
```
src/Text/Pandoc/Writers/Shared.hs:174:41: error:
    • Couldn't match expected type ‘a’ with actual type ‘Text’
      ‘a’ is a rigid type variable bound by
        the type signature for:
          tagWithAttrs :: forall a. HasChars a => Text -> Attr -> Doc a
        at src/Text/Pandoc/Writers/Shared.hs:173:1-51
    • In the first argument of ‘literal’, namely ‘tag’
      In the first argument of ‘(<>)’, namely ‘(literal tag)’
      In the second argument of ‘(<>)’, namely
        ‘(literal tag) <> (htmlAttrs attr) <> ">"’
    • Relevant bindings include
        tagWithAttrs :: Text -> Attr -> Doc a
          (bound at src/Text/Pandoc/Writers/Shared.hs:174:1)
    |
174 | tagWithAttrs tag attr = "<" <> (literal tag) <> (htmlAttrs attr) <> ">"
    |                                         ^^^
cabal: Failed to build pandoc-3.0 (which is required by exe:pandoc from
pandoc-cli-0.1).
```

I do not know enough about haskell to solve this.

Reply:
I see. Here's the type of `literal`

literal :: [HasChars](https://hackage.haskell.org/package/doclayout-0.4/docs/Text-DocLayout.html#t:HasChars) a => a -> [Doc](https://hackage.haskell.org/package/doclayout-0.4/docs/Text-DocLayout.html#t:Doc) a

So if `tag` is a `Text`, then `literal tag` will be a `Doc Text`. However, we've specified this function as polymorphic, it can produce `Doc a` for any `a` that is an instance of the `HasChars` class.  That's why it doesn't work, my mistake!

We could change the type of the function to make `tag` `a` rather than `Text`.  You could try that (and use `literal tag`), but I don't know if there are other places in the code that presuppose that tag is Text. I guess the compiler would let you know.

Reply:
Thank you for the explanation, I am starting to understand Haskell a lot more!
Changing the type definition of `tagWithAttrs` worked (in hindsight it is quite obvious haha).